programs/uxd/src/instructions/redeem_from_router.rs

```diff
@@ -0,0 +1,350 @@
+        token::authority = mercurial_vault_depository_0,
+        token::mint = mercurial_vault_depository_0_vault_lp_mint,
+        bump = mercurial_vault_depository_0.load()?.lp_token_vault_bump,
+    )]
+    pub mercurial_vault_depository_0_lp_token_vault: Box<Account<'info, TokenAccount>>,
+
+    /// #12
+    #[account(
+        mut,
+        constraint = mercurial_vault_depository_0_vault.token_vault == mercurial_vault_depository_0_collateral_token_safe.key() @UxdError::InvalidMercurialVaultCollateralTokenSafe,
+    )]
+    pub mercurial_vault_depository_0_vault: Box<Account<'info, mercurial_vault::state::Vault>>,
+
+    /// #13
+    #[account(mut)]
+    pub mercurial_vault_depository_0_vault_lp_mint: Box<Account<'info, Mint>>,
+
+    /// #14 - Token account owned by the mercurial vault program. Hold the collateral deposited in the mercurial vault.
+    #[account(mut)]
+    pub mercurial_vault_depository_0_collateral_token_safe: Box<Account<'info, TokenAccount>>,
+
+    /// #15
+    #[account(
+        mut,
+        seeds = [CREDIX_LP_DEPOSITORY_NAMESPACE, credix_lp_depository_0.load()?.credix_global_market_state.key().as_ref(), credix_lp_depository_0.load()?.collateral_mint.as_ref()],
+        bump = credix_lp_depository_0.load()?.bump,
+        has_one = controller @UxdError::InvalidController,
+        has_one = collateral_mint @UxdError::InvalidCollateralMint,
+    )]
+    pub credix_lp_depository_0: AccountLoader<'info, CredixLpDepository>,
+
+    /// #23
+    pub system_program: Program<'info, System>,
+
+    /// #24
+    pub token_program: Program<'info, Token>,
+
+    /// #25
+    pub associated_token_program: Program<'info, AssociatedToken>,
+
+    /// #26
+    pub mercurial_vault_program: Program<'info, mercurial_vault::program::Vault>,
+
+    /// #28
+    pub uxd_program: Program<'info, crate::program::Uxd>,
+
+    /// #29
+    pub rent: Sysvar<'info, Rent>,
+}
+
+struct DepositoryInfoForRedeemFromRouter<'info> {
+    pub weight_bps: u16,
+    pub redeemable_amount_under_management: u128,
+    pub redeemable_amount_under_management_cap: u128,
+    pub redeem_cpi: Option<Box<dyn Fn(u64) -> Result<()> + 'info>>,
+}
+
+pub(crate) fn handler(ctx: Context<RedeemFromRouter>, redeemable_amount: u64) -> Result<()> {
+    // Gather all the onchain states we need (caps, weights and supplies)
+    let controller = ctx.accounts.controller.load()?;
+    let identity_depository = ctx.accounts.identity_depository.load()?;
+    let mercurial_vault_depository_0 = ctx.accounts.mercurial_vault_depository_0.load()?;
+    let credix_lp_depository_0 = ctx.accounts.credix_lp_depository_0.load()?;
+
+    // The actual post-redeem circulating supply may be slightly higher
+    // Due to redeem fees and precision loss. But the difference should be negligible and
+    // Any future mint/redeem will recompute the targets based on the exact future circulating supply anyway
+    let mimimum_after_redeem_circulating_supply = controller
```

Comment:
At first glance I would say that:

In case of redeem, all the provided UXD are burnt, fees are paid by sending back less collateral to the user.

The calculated circulating supply here is correct.

Reply:
True true, but the precision loss may still impact it and the final circulating supply may be slightly above the computed amount here

Reply:
for example:

redeem 7units
- 3units from identity
- 3units from mercurial

total = 6 redeemed vs 7 estimated

Reply:
oh I see, meaning that the user may still have 1 native unit of UXD after redeeming.

should we burn that extra unit to make clean accounting for user? Or put that extra 1 native unit on whatever depository target.

Reply:
I think it can be sent to any depository (or identity if you wanna be safe)

Reply:
So actually the maths on this are not trivial for example:
- Lets say you have 1 or 2 native unit of error (from `total-SUM()`)
- You cant just do +1 to identity depository, otherwise if the identity depository is empty or at-limit it make the whole thing fail (since redeeming past 0 is an error)
- You would have to loop over all depository to find one that still has remaining redeemable after the regular redeem
- Since there could be 3 depositories redeemed from, but only one with remaining redeemable space (and it can be a different one than the one with the precision error)

Reply:
I don't think there is a shortcut that works on all cases here. A counter example would be:

Total Requested 11unit (bump +2 = 13):
 - Depository 1 (25%) -> 13*0.25 = 3
 - Depository 2 (25%) -> 13*0.25 = 3
 - Depository 2 (25%) -> 13*0.25 = 3
 - Depository 2 (25%) -> 13*0.25 = 3
Total Final: 12 - 2 = 10 which is different than initial 11    

Not ideal counter-example, but im pretty sure more such example can be found

Reply:
Yeah that could work, but its probably even more complicated than the actual fix probably, especially the random part. Also I don't know for sure this would work in all cases, as the weights and number of depositories can change 🤔

Reply:
> * You would have to loop over all depository to find one or two that still has remaining redeemable after the regular redeem

I think that would be the easiest safe fix that would work 100%, but unsure if its worth actually doing it ? I can give it a shot

Reply:
Oh wait, but if mercurial has precision loss, we wouldnt be able to account for that anyway 🤔

Reply:
What do you mean by that?

Any UXD you give to Mercurial or Credix depository are burned and user receives collateral. Collateral may suffer from precision loss, but it doesn't affect UXD circulating supply/user balance.

Reply:
Ah yes, you're right, struggling a bit to think straight. Let me try to edit `calculate_depositories_redeemable_amount` to try to make it have zero precision loss

Reply:
This should fix this issue here: https://github.com/UXDProtocol/uxd-program/pull/250/commits/a0bd6c43bfdc870cdb9c4b22ae441e20e1b95d28

Reply:
Lets continue the thread over there if we need 👌